File: src/TrainingLoadAnalyzer.Web/Features/Sync/SyncMessage.cs
using System.Globalization;
using TrainingLoadAnalyzer.Infrastructure.Sync;

namespace TrainingLoadAnalyzer.Web.Features.Sync;

/// <summary>
///   What the page says about the last sync (FR-010, US5 scenarios 1-5).
/// </summary>
/// <remarks>
///   A pure function of the status, so every one of US5's scenarios is a plain assertion with no
///   rendering and no Strava. Total by construction: the final arm of the switch means an outcome
///   added to feature 005 later still produces something a person can read, rather than an empty
///   panel (C97, Principle VI).
/// </remarks>
public static class SyncMessage
{
    /// <summary>The prompt that also appears in the empty state, so both routes read alike.</summary>
    public const string ConnectionRequired = "Strava connection required.";

    public static string For(SyncStatus status)
    {
        ArgumentNullException.ThrowIfNull(status);

        if (status.IsRunning)
        {
            return "Syncing activities…";
        }

        if (status.Failure is not null)
        {
            return ConnectionRequired;
        }

        if (status.Result is not { } result)
        {
            return string.Empty;
        }

        return result.Outcome switch
        {
            SyncOutcome.Completed when result.Imported > 0 =>
                string.Create(CultureInfo.InvariantCulture, $"{result.Imported} activities imported."),
            SyncOutcome.Completed => "Already up to date.",
            SyncOutcome.RateLimited => RateLimited(status),
            SyncOutcome.Interrupted => "Sync interrupted. Your stored history is unchanged — try again.",
            SyncOutcome.ReconnectionRequired => ConnectionRequired,
            SyncOutcome.Refused => "A sync is already running.",

            // Not a default that shrugs: an outcome nobody anticipated still has to say something,
            // because a blank panel after a click is indistinguishable from a broken page.
            _ => "Sync finished with an outcome this page does not recognise.",
        };
    }

    /// <summary>
    ///   005 FR-035: Strava's own next window boundary, not a fixed delay — so the athlete is told
    ///   when to come back rather than merely to wait. Shown in their local time, because a UTC
    ///   instant is not an answer to "when?".
    /// </summary>
    private static string RateLimited(SyncStatus status) =>
        status.RetryAfterLocal is { } retryAfter
            ? string.Create(
                CultureInfo.InvariantCulture,
                $"Rate limited by Strava. Available again at {retryAfter:HH:mm}.")
            : "Rate limited by Strava. Try again shortly.";
}

File: alt-stack/backend/src/tla/dashboard/sync_message.py
"""What the page says about the last sync: SyncMessage.cs, verbatim (006 FR-010, US5)."""

from typing import TYPE_CHECKING

from tla.sync.results import SyncOutcome

if TYPE_CHECKING:
    # Only for the annotation: the coordinator imports this module for its prompt.
    from tla.sync.coordinator import SyncStatus

# Also what the empty state prompts, so both routes read alike.
CONNECTION_REQUIRED = "Strava connection required."


def for_status(status: "SyncStatus") -> str:
    if status.is_running:
        return "Syncing activities…"
    if status.failure is not None:
        return CONNECTION_REQUIRED
    result = status.result
    if result is None:
        return ""

    match result.outcome:
        case SyncOutcome.COMPLETED if result.imported > 0:
            return f"{result.imported} activities imported."
        case SyncOutcome.COMPLETED:
            return "Already up to date."
        case SyncOutcome.RATE_LIMITED:
            retry = status.retry_after_local
            if retry is None:
                return "Rate limited by Strava. Try again shortly."
            return f"Rate limited by Strava. Available again at {retry.hour:02d}:{retry.minute:02d}."
        case SyncOutcome.INTERRUPTED:
            return "Sync interrupted. Your stored history is unchanged — try again."
        case SyncOutcome.RECONNECTION_REQUIRED:
            return CONNECTION_REQUIRED
        case SyncOutcome.REFUSED:
            return "A sync is already running."
        case _:
            # An outcome nobody anticipated still says something: a blank panel reads as broken.
            return "Sync finished with an outcome this page does not recognise."
